core/src/main/java/org/renjin/stats/internals/distributions/RNG.java

```diff
@@ -87,37 +88,57 @@ public static void set_seed(@Current Context context, int seed, SEXP kind, SEXP
   }
 
   @Internal
-  public static DoubleVector runif(@Current Context context, int n, double a, double b) {
+  public static DoubleVector runif(@Current Context context, Vector n, Vector a, Vector max) {
+    double min = a.getElementAsDouble(0);
+    int n1 = (n.length() != 1) ? n.length() : n.getElementAsInt(0);
     RNG rng = context.getSession().rng;
-    DoubleArrayVector.Builder vb = DoubleArrayVector.Builder.withInitialCapacity(n);
-    for (int i = 0; i < n; i++) {
-      vb.add(a + rng.unif_rand() * (b - a));
+    DoubleArrayVector.Builder vb = DoubleArrayVector.Builder.withInitialCapacity(n1);
+    int j = 0;
+    for (int i = 0; i < n1; i++) {
+      if (max.length() != 1 && j == max.length()-1) {
+        j = 0;
+      }
+      vb.add(min + rng.unif_rand() * (max.getElementAsDouble(j) - min));
+      j++;
     }
     return (vb.build());
   }
 
   @Internal
-  public static DoubleVector rnorm(@Current Context context, int n, double mean, double sd) {
-    DoubleArrayVector.Builder vb = DoubleArrayVector.Builder.withInitialCapacity(n);
-    for (int i = 0; i < n; i++) {
-      vb.add(Normal.rnorm(context.getSession(), mean, sd));
+  public static DoubleVector rnorm(@Current Context context, Vector n, double mean, Vector sd) {
+    int n1 = (n.length() != 1) ? n.length() : n.getElementAsInt(0);
+    DoubleArrayVector.Builder vb = DoubleArrayVector.Builder.withInitialCapacity(n1);
+    int j = 0;
+    for (int i = 0; i < n1; i++) {
+      if (sd.length() != 1 && j == sd.length()-1) {
+        j = 0;
+      }
+      vb.add(Normal.rnorm(context.getSession(), mean, sd.getElementAsDouble(j)));
+      j++;
     }
     return (vb.build());
   }
 
   @Internal
-  public static DoubleVector rgamma(@Current Context context, int n, double shape, double scale) {
-    DoubleArrayVector.Builder vb = DoubleArrayVector.Builder.withInitialCapacity(n);
-    for (int i = 0; i < n; i++) {
-      vb.add(Gamma.rgamma(context.getSession(), shape, scale));
+  public static DoubleVector rgamma(@Current Context context, Vector n, double shape, Vector scale) {
```

Comment:
Why are some parameters recycled and others not? Shouldn't `shape` also be recycled?